main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -313,25 +316,27 @@ object OccurrenceAnalyzer {
     case Expression.PutChannel(exp1, exp2, tpe, loc) =>
       val (e1, o1) = visitExp(exp1)
       val (e2, o2) = visitExp(exp2)
-      val o3 = combineAll(o1, o2)
+      val o3 = combineAllNonBranch(o1, o2)
       (OccurrenceAst.Expression.PutChannel(e1, e2, tpe, loc), o3)
 
     case Expression.SelectChannel(rules, default, tpe, loc) =>
       var rs: List[OccurrenceAst.SelectChannelRule] = List.empty
       var o1: Map[Symbol.VarSym, Occur] = Map.empty
+      var o2: Map[Symbol.VarSym, Occur] = Map.empty
       for (r <- rules) {
-        val (c, o2) = visitExp(r.chan)
-        val (e, o3) = visitExp(r.exp)
+        val (c, o3) = visitExp(r.chan)
+        val (e, o4) = visitExp(r.exp)
         rs = rs :+ OccurrenceAst.SelectChannelRule(r.sym, c, e)
-        o1 = combineAll(o1, combineAll(o2, o3))
+        o1 = combineAllNonBranch(o1,o3)
+        o2 = combineAllBranches(o2, o4)
       }
 
-      val (d1, o4) = default.fold[(Option[OccurrenceAst.Expression], Map[Symbol.VarSym, Occur])](None, o1)(x => {
-        val (d2, o5) = visitExp(x)
-        (Some(d2), combineAll(o5, o1))
+      val (d1, o5) = default.fold[(Option[OccurrenceAst.Expression], Map[Symbol.VarSym, Occur])](None, o1)(x => {
```

Comment:
This seems very complicated.

Reply:
Do you have a less complicated solution?

Reply:
Maybe with unzip